### ISSUE TITLE
A small trouble in R

### ISSUE THREAD

**Issue by kaleidoscopical** [dmlc/xgboost#236]
I have no idea of the occurring problem that by running the code below in R, the test samples with y1=0 is all fail to be correctly classified.

require(xgboost)
require(Matrix)
require(data.table)

x1 = seq(-5,5,0.01)
y1 = sin(x1)+runif(length(x1))
x2 = seq(-5,5,0.01)
y2 = sin(x2)-runif(length(x2))

data = rbind(cbind(x1,y1),cbind(x2,y2))
data_table = data.table(data, keep.rownames = F)
train_spar = sparse.model.matrix(~.,data = data_table)
train_label = cbind(c(rep(1,length(x1)),rep(0,length(x1))))
dtrain = xgb.DMatrix(data = train_spar,label = train_label)

xx = seq(-6,6,0.1)
yy = seq(-3,3,0.1)
zz = NULL
for(i in 1:length(xx))
    for(j in 1:length(yy))
        zz = rbind(zz,c(xx[i],yy[j]))

test_table = data.table(zz,keep.rownames=F)
test_spar = sparse.model.matrix(~.,data = test_table )
dtest = xgb.DMatrix(data = test_spar)

# begin to train the model
watchlist <- list(train=dtrain)
param <- list(max.depth = 2,gamma=1, eta = 0.1)
num_round <- 333
bst=xgb.train(param,dtrain,num_round,watchlist,objective = "binary:logistic")

preds = predict(bst,dtest)
pred = preds >= 0.5
pred = as.numeric(pred)

plot(x1,y1,col='red',xlim = c(-6,6), ylim = c(-3,3))
points(x2,y2,col='blue')

index_pos = which(pred==1)
index_neg = setdiff(1:length(pred),index_pos)
points(zz[index_pos,1],zz[index_pos,2],pch=19,cex=1,col='white')
points(zz[index_neg,1],zz[index_neg,2],pch=19,cex=1,col='grey')
points(x1,y1,col='red')
points(x2,y2,col='blue')

**Comment by kaleidoscopical** [comment]
could you please tell me where is wrong?

**Comment by tqchen** [comment]
It seems have things to do with generalization and model assumption.

In one dimension, the tree based model is basically trying to approximating the data with step function. 
On two dimension it was trying to do so by piece wise function. In general, this means that the model will output a constant  on the region it did not see.

On your example, you are testing on x that did not occurred on the training data, because tree have not see data in such x, it can only guess the region x > 5 maybe, and a constant value.

If you have prior knowledge of the function should be periodic, change input from x to (x MODE period) might resolve the problem

**Comment by tqchen** [comment]
I am closing this issue for now since i think the previous post answers the questions
